### ISSUE TITLE
.interactive dataframe slice selection by row index or position not possible

### ISSUE THREAD

**Issue by JanHomann** [holoviz/hvplot#771]
#### Is your feature request related to a problem? Please describe.
I was in the assumption that the .interactive interface for dataframes should mimic the interface of pandas. Yet not everything seems to work.

Given the code
```python
import hvplot.pandas
from bokeh.sampledata import airports
df = airports.data
dfi = df.interactive()

p = pnw.IntSlider(start=10, end=40)
dfi[5:p]
```
 I would like to see a slider and a dynamic dataframe representation showing index 5...p-1 (that dataframe has an integer index). Instead I get a static output without slider. I can make one in another cell, but when I move it, the dataframe output does not update. I need to execute the cell manually again to get the update. 

This here doesn't work at all and gives an InvalidIndexError
```python
dfi.loc[5:p,:]
# InvalidIndexError: IntSlider(end=40, start=10, value=10)
```

Those here give no slider and don't update when I create a slider in another cell:
```python
dfi.loc[:p]

dfi.iloc[5:p]
dfi.iloc[slice(5,p)]
dfi.values[5:p]
```

This here gives a type error
```python
dfi.loc[dfi.index[5:p]]
# TypeError: slice indices must be integers or None or have an __index__ method
```

A current workaround for `dfi.loc[5:p,:]` is:
```python
dfi[(dfi.index>=5) & (dfi.index<p)]
```

For `dfi.loc[5:p,:]` I have no reasonable workaround (except for resetting the index, selecting via above method and setting it again).

Given that neither `dfi[:p]` nor `dfi.iloc[:p,:]` nor `dfi.values[5:p]` nor `dfi.loc[dfi.index[5:p]]` work, I don't know how I would even select rows by position. 
Literally the first example in the interactive tutorial (https://hvplot.holoviz.org/user_guide/Interactive.html) has indexing by position in an xarray with `.isel`, which I think i have tried and it works. So it is surprising to me that there is no mechanism to select a positional range in dataframes. Or maybe there is some Pandas trick that I don't know.

**Comment by JanHomann** [comment]
Yes, I suppose the mechanism by which this works in Pandas is that a slice object is passed to the dataframe method. Makes sense. Is there currently a workaround for positional indexing?

**Comment by JanHomann** [comment]
Here is somewhat of a workaround. Unfortunately that only works when you start out with a dataframe, not with an interactive dataframe. So essentially if you know that you need positional indexing in your interactive dataframe, you have to do it right when creating it from the normal dataframe if you use this method. Passing an interactive dataframe into this mechanism (like with `df=dfi`) doesn't work.

```python
end_ind = pnw.IntSlider(start=10, end=40)

def make_slice(df=df, end_ind=20):
    return df.iloc[5:end_ind]

dfi = hvplot.bind(make_slice, df=df, end_ind=end_ind).interactive(width=600)
dfi   # works as expected with dynamic slider
```

**Comment by Hoxbro** [comment]
I got it to work with slices, and will submit PRs in holoviews and hvplot today.

https://user-images.githubusercontent.com/19758978/177975269-75bdf24f-7fee-4f08-a17d-906b0a5cf9d6.mp4

**Comment by JanHomann** [comment]
Wow, you guys are amazing. A solution in less than 24 hours?!

**Comment by jbednar** [comment]
Nice! @JanHomann , a really valuable contribution you (or other users) could make is to study the Pandas API and see if there are any other collections or special objects that would need similar treatment. I think we current now handle slices, dicts, and lists; not sure if there special iterators or other objects that we should be looking out for...

**Comment by JanHomann** [comment]
> Nice! @JanHomann , a really valuable contribution you (or other users) could make is to study the Pandas API and see if there are any other collections or special objects that would need similar treatment. I think we current now handle slices, dicts, and lists; not sure if there special iterators or other objects that we should be looking out for...

Thank you! Another thing I have noticed is that dataframe functions don't give back the correct type. For example:

```python
p = pnw.IndexSlider()
df.columns(p)
```
doesn't return a string (the column name), but it returns another interactive object, that cannot be used for example for indexing. So:

```python
df[df.columns[p]]
```
doesn't work because it gets the wrong type.

**Comment by JanHomann** [comment]
Here is another function that I use quite a lot: The query function. I think this function can be replicated by normal boolean selection, but it's generally more concise and speed-wise pretty good. 

```python
import hvplot.pandas
import panel.widgets as pnw
from bokeh.sampledata import antibiotics

df = antibiotics.data
dfi = df.interactive()

p = pnw.IntSlider(value=5, start=1, end=10)
df.query('penicillin < @p')

# TypeError
```

**Comment by JanHomann** [comment]
And then there is this:

It's not a query on dfi. So probably it's not intended to work.

<img width="615" alt="Screen Shot 2022-07-08 at 6 01 38 PM" src="https://user-images.githubusercontent.com/35338267/178075785-52b0da1b-af79-4576-a0ed-ab5d99fb90d1.png">

**Comment by JanHomann** [comment]
Overall I totally love the holoviz stack. I think it's the most advanced and user friendly stack for interactive plotting.

**Comment by JanHomann** [comment]
And then there are of course (lambda) function that some of the pandas methods take. 

```python
import pandas as pd
import hvplot.pandas


df = pd.DataFrame(data=np.random.randn(10,3))
p = pnw.IntSlider(start=1, end=10, value=5)
dfi = df.interactive()

df.apply(lambda x: x*5)   # this works
dfi.apply(lambda x: x*p)  # this doesn't
```

**Comment by JanHomann** [comment]
Range objects can also be passed to some function.

```python
import numpy as np
import pandas as pd
import hvplot.pandas
import panel.widgets as pnw


df = pd.DataFrame(data=np.random.randn(10,3))
p = pnw.IntSlider(start=1, end=10, value=5)
dfi = df.interactive()

df.loc[range(1,10),:]     # this works
dfi.loc[range(1,p),:]     # this doesn't
# TypeError: 'IntSlider' object cannot be interpreted as an integer

**Comment by JanHomann** [comment]
Generator objects can also be passed to some pandas functions (I think essentially everything that would takes a list.).

```python
import numpy as np
import pandas as pd
import hvplot.pandas
import panel.widgets as pnw

df = pd.DataFrame(data=np.random.randn(20,3))
dfi = df.interactive()
p = pnw.IntSlider(start=1, end=4, value=2)
g = (p**2 for n in range(5))
g 1 = (n**2 for n in range(p))               # panel widget IntSlider is embedded in a generator

df.loc[g,:]            # this works

g1 = (n**2 for n in range(p))
dfi.loc[g1,:]           # this one doesn't:    TypeError: 'IntSlider' object cannot be interpreted as an integer

**Comment by JanHomann** [comment]
So we have currently not supported by .interactive but by pandas for certain functions: 

- f-strings, 
- normal strings with @parameter for .query and .eval
- range objects
- (lambda) functions
- generators

**Comment by JanHomann** [comment]
Another thing that doesn't work having a slider in a `pd.Timestamp` object, which can also be something that is understood by Pandas as input for a computation. I bet a slider in a `pd.TimeDelta` doesn't work either. 

```python
from bokeh.sampledata import daylight
df = daylight.daylight_warsaw_2013
dfi = df.interactive()

df[df.Date < pd.Timestamp(year=2013, month=6, day=5)]     # This works
dfi[dfi.Date < pd.Timestamp(year=2013, month=6, day=p)]   # this one doesn't
#TypeError: an integer is required (got type IntSlider)
```

**Comment by JanHomann** [comment]
I think a few Pandas functions also can take an `pd.Interval` object as an input. Intervals can be used for example as elements of an index (`pd.Intervalndex`)and then a `pd.Interval` object can be used to filter down the IntervalIndex for rows that overlap with that interval. Pandas has rather mediocre support for intervals, but for example `pd.cut()` (used for binning continuous data) returns intervals that then can be used by with `.groupby()` to get statistics on those bins.

```python
df = pd.DataFrame(data=np.random.randn(6,3), 
                  index=pd.IntervalIndex.from_tuples( [ (-1.5,1), (-1,-0.5), (-0.5,0), (0,0.5), (0.5,1), (1,1.5) ]
                                                     )
                  )

dfi = df.interactive()
p = pnw.IntSlider(start=1, end=3, value=2)

df.index.overlaps(pd.Interval(0.2,1))    # this one works (returns a boolean array)
dfi.index.overlaps(pd.Interval(0.2,p))   # this one doesn't 
# ValueError: Only numeric, Timestamp and Timedelta endpoints are allowed when constructing an Interval.

df.index < 2                  # this one gives back a boolean array
dfi.index < p                 # this one doesn't work either (no error message, but it gives back a slider and an empty list)

```

**Comment by JanHomann** [comment]
So in summary,
 currently unsupported are:

- f-strings
- embedding variables with @ for .query and .eval
- range objects
- (lambda) functions
- generators
- pd.Timestamp
- pd.TimeDelta
- pd.Interval
- item assignment
- feeding the output of an interactive dataframe into an interactive dataframe method.

**Comment by Hoxbro** [comment]
Great work @JanHomann! I will look into what is possible to add to interactive.

**Comment by JanHomann** [comment]
@Hoxbro The original problem seems solved now. So should this stay open? Or should the title be renamed, because we figured that there are other DataFrame parameter types that are currently not supported by interactive dataframes?

**Comment by Hoxbro** [comment]
Let's keep this open with a title rename. 

We have begun rewriting `.interactive` to be more generic. When that is in place, we will revisit the suggestion made in this thread.
